File: src/supervisor.cpp
#include <ros/ros.h>
#include <geometry_msgs/PoseStamped.h>
#include <nav_msgs/Odometry.h>
#include <sensor_msgs/NavSatFix.h>
#include <geographic_msgs/GeoPointStamped.h>
#include <geometry_msgs/PoseArray.h>
#include <geometry_msgs/Pose.h>
#include <tf2/LinearMath/Quaternion.h>
#include <tf2_ros/transform_broadcaster.h>
#include <sensor_msgs/Range.h>
#include <geometry_msgs/TransformStamped.h>

#include <cmath>
#include <Eigen/Dense>
#include <Eigen/Core>
#include <vector>
#include <math.h>
#include <chrono>
#include <signal.h>
#include <tf2/LinearMath/Quaternion.h>
#include <tf2/utils.h>


#define M_PI   3.14159265358979323846  /*pi*/

using namespace std::chrono_literals;
using std::placeholders::_1;

sig_atomic_t volatile node_shutdown_request = 0;    //signal manually generated when ctrl+c is pressed


// Global vars
double EARTH_RADIUS = 6378137.0;

Eigen::Vector3d global_to_local(double lat, double lon, double alt,
                                double origin_lat, double origin_lon, double origin_alt)
{
    double d_lat = (lat - origin_lat) * M_PI / 180.0;
    double d_lon = (lon - origin_lon) * M_PI / 180.0;
    
    double lat1 = origin_lat * M_PI / 180.0;
    double lat2 = lat * M_PI / 180.0;

    Eigen::Vector3d local_pos;
    local_pos(0) = EARTH_RADIUS * d_lon * cos((lat1 + lat2) / 2);
    local_pos(1) = EARTH_RADIUS * d_lat;
    local_pos(2) = alt - origin_alt;

    return local_pos;
    
}



class Supervisor
{
    public:
        Supervisor() : nh_priv_("~")
        {
            // ---------------- ROS params -----------------
            this->nh_.getParam("robots_ids", IDS);
            std::cout << "Got IDS: ";
            for (auto id : IDS) {
                std::cout << id << ", ";
            }
            std::cout << "\n";
            ROBOTS_NUM = IDS.size();
            std::cout << "Number of robots : " << ROBOTS_NUM << std::endl;
            std::cout << "Using home position of robot " << IDS[0] << std::endl;

            // -------------- Subs and Pubs ----------------
            gps_subs_.resize(ROBOTS_NUM);
            odom_subs_.resize(ROBOTS_NUM);
            odom_pubs_.resize(ROBOTS_NUM);

            // Origin is starting position of UAV0 for every UAV
            // home_sub_ = nh_.subscribe<geographic_msgs::GeoPointStamped>("/uav"+std::to_string(IDS[0])+"/mavros/global_position/gp_origin", 1, std::bind(&Supervisor::home_callback, this, std::placeholders::_1));
            home_sub_ = nh_.subscribe<sensor_msgs::NavSatFix>("/uav"+std::to_string(IDS[0])+"/mavros/global_position/global", 1, std::bind(&Supervisor::home_callback, this, std::placeholders::_1));
            for (int i = 0; i < ROBOTS_NUM; ++i)
            {
                gps_subs_[i] = nh_.subscribe<sensor_msgs::NavSatFix>("/uav"+std::to_string(IDS[i])+"/mavros/global_position/global", 1, std::bind(&Supervisor::gps_callback, this, std::placeholders::_1, i));
                odom_subs_[i] = nh_.subscribe<nav_msgs::Odometry>("/uav"+std::to_string(IDS[i])+"/mavros/local_position/odom", 1, std::bind(&Supervisor::odom_callback, this, std::placeholders::_1, i));
                odom_pubs_[i] = nh_.advertise<nav_msgs::Odometry>("/supervisor/uav"+std::to_string(IDS[i])+"/odom", 10);
            }
            timer_ = nh_.createTimer(ros::Duration(0.01), std::bind(&Supervisor::timer_callback, this));



            // Init robot poses (each robot in a column)
            robots.resize(13, ROBOTS_NUM);       // x,y,z,rx,ry,rz,rw,vx,vy,vz,vrx,vry,vrz
            robots.setZero();
            robots.row(6).setOnes();            // consistent quaternion
            starting_poses.resize(3, ROBOTS_NUM);       // x,y,z
            starting_poses.setZero();
            robots_local.resize(13, ROBOTS_NUM);
            robots_local.setZero();
            robots_local.row(6).setOnes();            // consistent quaternion
            p_i.resize(7);
            p_i.setZero();
            p_i(6) = 1;               // consistent quaternion
            R_i.setZero();

            got_starting_poses.resize(ROBOTS_NUM, false);

            
            /*transform.setOrigin(tf::Vector3(robots(0, ROBOT_ID), 
                                    robots(1, ROBOT_ID), 
                                    robots(2, ROBOT_ID)));

            transform.setRotation(tf::Quaternion(robots(3, ROBOT_ID),
                                        robots(4, ROBOT_ID),
                                        robots(5, ROBOT_ID),
                                        robots(6, ROBOT_ID)));
            br.sendTransform(tf::StampedTransform(transform, ros::Time::now(), "map", "uav_"+std::to_string(ROBOT_ID)));
            */

            std::cout << "Init complete.\n";

        } 

        ~Supervisor()
        {
            std::cout << "Supervisor destroyer called\n";
        }

        void stop();
        void timer_callback();
        void emulate_vision();
        void home_callback(const sensor_msgs::NavSatFix::ConstPtr& msg);
        void gps_callback(const sensor_msgs::NavSatFix::ConstPtr& msg, int id);
        void odom_callback(const nav_msgs::Odometry::ConstPtr& msg, int id);

    private:
        std::vector<int> IDS = {0};
        int ROBOTS_NUM;
        bool got_home = false;

        // Global origin
        double origin_lat = 0.0;
        double origin_lon = 0.0;
        double origin_alt = 0.0;
        
        // Robots poses
        Eigen::MatrixXd robots;
        Eigen::MatrixXd starting_poses;
        Eigen::MatrixXd robots_local;
        Eigen::VectorXd p_i;                // UAV[i] pose
        Eigen::Matrix3d R_i;                // Rotation matrix 
        std::vector<bool> got_starting_poses;

        // ROS
        ros::NodeHandle nh_;
        ros::NodeHandle nh_priv_;
        ros::Timer timer_;
        ros::Subscriber home_sub_;
        std::vector<ros::Subscriber> gps_subs_;
        std::vector<ros::Subscriber> odom_subs_;
        std::vector<ros::Publisher> odom_pubs_;

        tf2_ros::TransformBroadcaster br;


};

void Supervisor::stop()
{
    ros::Duration(0.1).sleep();
    ros::shutdown();
}

// void Supervisor::home_callback(const geographic_msgs::GeoPointStamped::ConstPtr& msg)
// {
//     origin_lat = msg->position.latitude;
//     origin_lon = msg->position.longitude;
//     origin_alt = msg->position.altitude;
//     got_home = true;
// }

void Supervisor::home_callback(const sensor_msgs::NavSatFix::ConstPtr& msg)
{
    if (!got_home) {
        origin_lat = msg->latitude;
        origin_lon = msg->longitude;
        origin_alt = msg->altitude;
        got_home = true;
    }
}


/* ---------- USE GPS FOR POSITION, ODOM FOR ORIENTATION AND VELOCITY ------*/
void Supervisor::gps_callback(const sensor_msgs::NavSatFix::ConstPtr& msg, int id)
{
    robots.col(id).head(3) = global_to_local(msg->latitude, msg->longitude, msg->altitude,
                                            origin_lat, origin_lon, origin_alt);
    if(!got_starting_poses[id]) {
        starting_poses.col(id) = robots.col(id).head(3);
        std::cout << "Starting position of robot " << IDS[id] << ": " << starting_poses.col(id).transpose() << std::endl;
        got_starting_poses[id] = true;
    }
    
}

void Supervisor::odom_callback(const nav_msgs::Odometry::ConstPtr& msg, int id)
{
    robots(3, id) = msg->pose.pose.orientation.x;
    robots(4, id) = msg->pose.pose.orientation.y;
    robots(5, id) = msg->pose.pose.orientation.z;
    robots(6, id) = msg->pose.pose.orientation.w;
    robots(7, id) = msg->twist.twist.linear.x;
    robots(8, id) = msg->twist.twist.linear.y;
    robots(9, id) = msg->twist.twist.linear.z;
    robots(10, id) = msg->twist.twist.angular.x;
    robots(11, id) = msg->twist.twist.angular.y;
    robots(12, id) = msg->twist.twist.angular.z;

    robots_local(0, id) = msg->pose.pose.position.x;
    robots_local(1, id) = msg->pose.pose.position.y;
    robots_local(2, id) = msg->pose.pose.position.z;
    robots_local(3, id) = msg->pose.pose.orientation.x;
    robots_local(4, id) = msg->pose.pose.orientation.y;
    robots_local(5, id) = msg->pose.pose.orientation.z;
    robots_local(6, id) = msg->pose.pose.orientation.w;
    robots_local(7, id) = msg->twist.twist.linear.x;
    robots_local(8, id) = msg->twist.twist.linear.y;
    robots_local(9, id) = msg->twist.twist.linear.z;
    robots_local(10, id) = msg->twist.twist.angular.x;
    robots_local(11, id) = msg->twist.twist.angular.y;
    robots_local(12, id) = msg->twist.twist.angular.z;
}

/*
void Supervisor::emulate_vision()
{
    p_i = robots.col(ROBOT_ID);
    tf2::Quaternion q(p_i(3), p_i(4), p_i(5), p_i(6));
    tf2::Matrix3x3 m(q);
    double roll, pitch, yaw;
    m.getRPY(roll, pitch, yaw);

    R_i(0, 0) = cos(yaw);
    R_i(0, 1) = sin(yaw);
    R_i(0, 2) = 0.0;
    R_i(1, 0) = -sin(yaw);
    R_i(1, 1) = cos(yaw);
    R_i(1, 2) = 0.0;
    R_i(2, 0) = 0.0;
    R_i(2, 1) = 0.0;
    R_i(2, 2) = 1.0;
    
    for (int i = 0; i < ROBOTS_NUM; ++i)
    {
        robots_local.col(i) = R_i * (robots.col(i).head(3) - p_i.head(3));
        // Simulate vision detection
        double dist = robots_local.col(i).norm();
        double angle = abs(atan2(robots_local(1, i), robots_local(0,i)));
        if (angle <= 0.5*ROBOT_FOV_rad && dist <= ROBOT_RANGE)
        {
            // Detected
            n_msg.poses[i].position.x = robots_local(0, i);
            n_msg.poses[i].position.y = robots_local(1, i);
            n_msg.poses[i].position.z = robots_local(2, i);

            if (i != ROBOT_ID)
            {
                geometry_msgs::Pose pose_msg;
                pose_msg.position.x = robots_local(0, i);
                pose_msg.position.y = robots_local(1, i);
                pose_msg.position.z = robots_local(2, i);
                target_pubs_[i].publish(pose_msg);

                // TF
                transformStamped.header.stamp = ros::Time::now();
                transformStamped.header.frame_id = "uav_"+std::to_string(ROBOT_ID);
                transformStamped.child_frame_id = "uav_"+std::to_string(ROBOT_ID)+"/target_"+std::to_string(i);
                transformStamped.transform.translation.x = robots_local(0, i);
                transformStamped.transform.translation.y = robots_local(1, i);
                transformStamped.transform.translation.z = robots_local(2, i);
                transformStamped.transform.rotation.x = 0.0;
                transformStamped.transform.rotation.y = 0.0;
                transformStamped.transform.rotation.z = 0.0;
                transformStamped.transform.rotation.w = 1.0;
                br.sendTransform(transformStamped);
            }
        }
    }
    n_msg.header.stamp = ros::Time::now();
    neigh_pub_.publish(n_msg);
}
*/

void Supervisor::timer_callback()
{
    if (!got_home) {
        std::cout << "Waiting for home ...\n";
        return;
    }


    for (int i = 0; i < ROBOTS_NUM; ++i) {

        // Publish local odom data
        nav_msgs::Odometry msg;
        msg.header.frame_id = "map";
        msg.header.stamp = ros::Time::now();
        // msg.child_frame_id = "uav" + std::to_string(IDS[i])+"/base_link";
        msg.pose.pose.position.x = robots(0, i);   
        msg.pose.pose.position.y = robots(1, i);   
        msg.pose.pose.position.z = robots(2, i);
        msg.pose.pose.orientation.x = robots(3, i);   
        msg.pose.pose.orientation.y = robots(4, i);   
        msg.pose.pose.orientation.z = robots(5, i);   
        msg.pose.pose.orientation.w = robots(6, i);
        msg.twist.twist.linear.x = robots(7, i);
        msg.twist.twist.linear.y = robots(8, i);
        msg.twist.twist.linear.z = robots(9, i);
        msg.twist.twist.angular.x = robots(10, i);
        msg.twist.twist.angular.y = robots(11, i);
        msg.twist.twist.angular.z = robots(12, i);
        odom_pubs_[i].publish(msg);

        // Publish TF
        
        geometry_msgs::TransformStamped ts;
        ts.header.stamp = ros::Time::now();
        ts.header.frame_id = "map";
        ts.child_frame_id = "uav" + std::to_string(IDS[i]) + "/odom";
        ts.transform.translation.x = starting_poses(0, i);
        ts.transform.translation.y = starting_poses(1, i);
        ts.transform.translation.z = starting_poses(2, i);
        ts.transform.rotation.x = 0.0;
        ts.transform.rotation.y = 0.0;
        ts.transform.rotation.z = 0.0;
        ts.transform.rotation.w = 1.0;
        br.sendTransform(ts);



        ts.header.stamp = ros::Time::now();
        ts.header.frame_id = "uav" + std::to_string(IDS[i])+"/odom";
        ts.child_frame_id = "uav" + std::to_string(IDS[i])+"/base_link";
        ts.transform.translation.x = robots_local(0, i);
        ts.transform.translation.y = robots_local(1, i);
        ts.transform.translation.z = robots_local(2, i);
        ts.transform.rotation.x = robots_local(3, i);
        ts.transform.rotation.y = robots_local(4, i);
        ts.transform.rotation.z = robots_local(5, i);
        ts.transform.rotation.w = robots_local(6, i);

        br.sendTransform(ts);
    }
    
    

}





/*******************************************************************************
* Main function
*******************************************************************************/
//alternatively to a global variable to have access to the method you can make STATIC the class method interested, 
//but some class function may not be accessed: "this->" method cannot be used

void nodeobj_wrapper_function(int){
    ROS_WARN("signal handler function CALLED");
    node_shutdown_request = 1;
}

int main(int argc, char* argv[])
{
    ros::init(argc, argv, "supervisor_node", ros::init_options::NoSigintHandler);
    signal(SIGINT, nodeobj_wrapper_function);

    //Controller node_controller;
    auto node_controller = std::make_shared<Supervisor>();

    while (!node_shutdown_request){
        ros::spinOnce();
    }
    node_controller->stop();

    //ros::spin();
    //do pre-shutdown tasks
    if (ros::ok())
    {
        ROS_WARN("ROS HAS NOT BEEN PROPERLY SHUTDOWN, it is being shutdown again.");
        ros::shutdown();
    }

    return 0;
}

File: src/supervisor.py
#!/usr/bin/env python3

import rospy
import tf2_ros
import numpy as np
from sensor_msgs.msg import NavSatFix
from nav_msgs.msg import Odometry
from geometry_msgs.msg import TransformStamped
from pyproj import Proj
import tf
import tf_conversions


class Supervisor():
    def __init__(self):
        rospy.init_node("tf_broadcaster")
        self.ids = rospy.get_param("robots_ids", "[0, 1]")
        self.offset_x = rospy.get_param("~offset_x", 0.0)
        self.offset_y = rospy.get_param("~offset_y", 0.0)
        self.offset_z = rospy.get_param("~offset_z", 0.0)
        self.rotation = rospy.get_param("~rotation", 0.0)
        self.rotation *= np.pi / 180.0
        self.R = np.array([
                [np.cos(self.rotation), -np.sin(self.rotation), 0.0],
                [np.sin(self.rotation), np.cos(self.rotation),  0.0],
                [0.0,                   0.0,                    1.0]])
        T = np.eye(4)
        T[0:3, 0:3] = self.R
        T[0:3, 3] = np.array([self.offset_x, self.offset_y, 0.0])
        self.quat = tf_conversions.transformations.quaternion_from_matrix(T)
        print("Using offsets: ", self.offset_x, self.offset_y, self.offset_z)
        print("Rotation: ", self.rotation)
        print("Got IDS:")
        for id in self.ids:
            print(f"{id}, ")

        self.robots_num = len(self.ids)
        print("Number of drones: ", self.robots_num)
        print("Using home position of robot ", self.ids[0])

        self.ref_lat = None
        self.ref_lon = None
        self.ref_alt = None
        self.proj = None
        self.gps_data = [None] * self.robots_num
        self.local_odom = [None] * self.robots_num
        self.just_init = [True] * self.robots_num
        self.starting_positions = np.zeros((self.robots_num, 3))

        self.tf_broadcaster = tf2_ros.TransformBroadcaster()
        self.static_br = tf2_ros.StaticTransformBroadcaster()
        self.global_odom_pubs = [None] * self.robots_num

        # Define pubs/subs
        for i, id in enumerate(self.ids):
            gps_topic = f"/uav{id}/mavros/global_position/global"
            rospy.Subscriber(gps_topic, NavSatFix, self.gps_callback, callback_args=i)
            odom_topic = f"/uav{id}/mavros/local_position/odom"
            rospy.Subscriber(odom_topic, Odometry, self.odom_callback, callback_args=i)
            global_odom_topic = f"/supervisor/uav{id}/odom"
            self.global_odom_pubs[i] = rospy.Publisher(global_odom_topic, Odometry, queue_size=1)

    def gps_callback(self, msg, i):
        self.gps_data[i] = msg

        # set reference origin
        if self.proj is None and i == 0:
            self.ref_lat = msg.latitude
            self.ref_lon = msg.longitude
            self.ref_alt = msg.altitude
            self.proj = Proj(proj='aeqd', 
                            lat_0=self.ref_lat,
                            lon_0=self.ref_lon,
                            ellps='WGS84')
            rospy.loginfo(f"[TF] Reference origin set from uav {self.ids[i]}: ({self.ref_lat}, {self.ref_lon}, {self.ref_alt})")

        # send map -> local origin (just once)
        if self.proj is not None:
            if self.just_init[i]:
                x, y = self.proj(msg.longitude, msg.latitude)
                z = msg.altitude - self.ref_alt
                self.starting_positions[i, 0] = x + self.offset_x
                self.starting_positions[i, 1] = y + self.offset_y
                self.starting_positions[i, 2] = z + self.offset_z
                # self.starting_positions[i, 2] = 0.0
                print(f"Starting position of uav {self.ids[i]}: {self.starting_positions[i]}")
                self.just_init[i] = False


    def odom_callback(self, msg, i):
        print(f"Odom for robot {self.ids[i]}: {msg.pose.pose.position}")
        self.local_odom[i] = msg

    def run(self):
        rate = rospy.Rate(20.0)
        while not rospy.is_shutdown():
            t_now = rospy.Time.now()
            if self.proj is None:
                rate.sleep()
                continue
            
            for i in range(self.robots_num):
                # [TF] map -> local origin
                tf_origin = TransformStamped()
                tf_origin.header.stamp = t_now
                tf_origin.header.frame_id = f"map"
                tf_origin.child_frame_id = f"uav{self.ids[i]}/local_origin"
                tf_origin.transform.translation.x = self.starting_positions[i, 0]
                tf_origin.transform.translation.y = self.starting_positions[i, 1]
                tf_origin.transform.translation.z = self.starting_positions[i, 2]
                # self.quat = tf_conversions.transformations.quaternion_from_euler(0, 0, self.rotation)
                # tf_origin.transform.rotation.x = self.quat[0]
                # tf_origin.transform.rotation.y = self.quat[1]
                # tf_origin.transform.rotation.z = self.quat[2]
                # tf_origin.transform.rotation.w = self.quat[3]
                tf_origin.transform.rotation.w = 1
                self.tf_broadcaster.sendTransform(tf_origin)
                
                odom = self.local_odom[i]
                #print(f"Odom for robot {i}: {odom.pose.pose.position}")
                if odom is not None:
                    # [TF] local origin -> base link
                    tf_odom = TransformStamped()
                    tf_odom.header.stamp = t_now
                    tf_odom.header.frame_id = f"uav{self.ids[i]}/local_origin"
                    tf_odom.child_frame_id = f"uav{self.ids[i]}/base_link"

                    tf_odom.transform.translation.x = odom.pose.pose.position.x
                    tf_odom.transform.translation.y = odom.pose.pose.position.y
                    tf_odom.transform.translation.z = odom.pose.pose.position.z
    
                    tf_odom.transform.rotation = odom.pose.pose.orientation

                    self.tf_broadcaster.sendTransform(tf_odom)

                    # [Topic] map -> base link
                    odom_msg = Odometry()
                    odom_msg.header.stamp = t_now
                    odom_msg.header.frame_id = "map"
                    odom_msg.pose.pose.position.x = odom.pose.pose.position.x + self.starting_positions[i, 0]
                    odom_msg.pose.pose.position.y = odom.pose.pose.position.y + self.starting_positions[i, 1]
                    odom_msg.pose.pose.position.z = odom.pose.pose.position.z + self.starting_positions[i, 2]
                    odom_msg.pose.pose.orientation = odom.pose.pose.orientation
                    self.global_odom_pubs[i].publish(odom_msg)



            rate.sleep()

if __name__ == "__main__":
    node = Supervisor()
    node.run()
